react-intl/index.d.ts

```diff
@@ -14,129 +12,110 @@ declare namespace ReactIntl {
+        withRef?: boolean;
+    }
 
-    function hasLocaleData(localeName: string): boolean;
+    type ComponentConstructor<P> = React.ComponentClass<P> | React.StatelessComponent<P>
+
+    function injectIntl<P>(component: ComponentConstructor<P & InjectedIntlProps>, options?: InjectIntlConfig): React.ComponentClass<P & InjectedIntlProps>;
+
+    function addLocaleData(data: Locale[] | Locale): void;
 
     interface Messages {
         [key: string]: FormattedMessage.MessageDescriptor
     }
 
     function defineMessages<T extends Messages>(messages: T): T;
 
-    interface IntlShape extends React.Requireable<any> {
+    interface IntlConfig {
+        locale: React.Requireable<any>;
+        formats: React.Requireable<any>;
+        messages: React.Requireable<any>;
+        defaultLocale: React.Requireable<any>;
+        defaultFormats: React.Requireable<any>;
     }
 
-    var intlShape: IntlShape;
+    interface IntlFormat {
+        formatDate: React.Requireable<any>;
+        formatTime: React.Requireable<any>;
+        formatRelative: React.Requireable<any>;
+        formatNumber: React.Requireable<any>;
+        formatPlural: React.Requireable<any>;
+        formatMessage: React.Requireable<any>;
+        formatHTMLMessage: React.Requireable<any>;
+    }
 
-    interface FormatConfig {
-        locale?: string;
-        formats?: any;
+    interface IntlShape extends IntlConfig, IntlFormat, React.Requireable<any> {
+        now: React.Requireable<any>;
     }
 
+    const intlShape: IntlShape;
+
     interface InjectedIntl {
         formatDate: (date: Date, options?: FormattedDate.PropsBase) => string;
         formatTime: (date: Date, options?: FormattedTime.PropsBase) => string;
-        formatRelative: (value: number, options?: FormattedRelative.PropsBase) => string;
+        formatRelative: (value: number, options?: FormattedRelative.PropsBase & { now?: any }) => string;
         formatNumber: (value: number, options?: FormattedNumber.PropsBase) => string;
-        formatPlural: (value: number, options?: FormattedPlural.PropsBase) => string;
-        formatMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: Object) => string;
-        formatHTMLMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: Object) => string;
+        formatPlural: (value: number, options?: FormattedPlural.Base) => keyof FormattedPlural.PropsBase;
+        formatMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: any) => string;
+        formatHTMLMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: any) => string;
+    }
+
+    interface InjectedConfig {
+        locale: string;
+        formats: any;
+        messages: { [id: string]: string };
+        defaultLocale: string;
+        defaultFormats: any;
     }
 
     interface InjectedIntlProps {
-        intl: InjectedIntl
+        intl: InjectedIntl & InjectedConfig & { now : () => number };
```

Comment:
Well in this case we can define separate type for each key of this object and say that it makes code clean. Each type is unnecessary abstraction which makes understanding and subsequent changes more complex. So my vote would be to have one type instead even if it is internal thing.